File: _exercises/13_plot-image-2/13_plot-image-2.R
library(readr)
library(ellmer)

m <- 32
u <- (seq_len(floor(sqrt(m))) - 0.5) / floor(sqrt(m))
grid <- as.matrix(expand.grid(x = u, y = u))

eps <- 1 / (2 * sqrt(m))
jitter <- matrix(runif(length(grid), -eps, eps), ncol = 2)
grid_jitter <- pmin(pmax(grid + jitter, 0), 1)

plot(
  x = grid_jitter[, 1],
  y = grid_jitter[, 2],
  pch = 19, # filled circles
  col = "steelblue",
  cex = 1.1,
  xlab = "Weight (1000 lb)",
  ylab = "Miles per Gallon (mpg)",
  main = "MPG vs Weight"
)

chat <- chat("openai/gpt-5", echo = "output")
chat$chat(
  "Interpret this plot.",
  content_image_plot()
)

File: _exercises/13_plot-image-2/13_plot-image-2.py
# %%
from math import floor, sqrt

import chatlas
import dotenv
import matplotlib.pyplot as plt
import numpy as np

dotenv.load_dotenv()

# %%
m = 32

g = floor(sqrt(m))
u = (np.arange(1, g + 1) - 0.5) / g
xx, yy = np.meshgrid(u, u)
grid = np.column_stack([xx.ravel(), yy.ravel()])

# small jitter to avoid perfect lattice, scaled to cell size
eps = 1.0 / (2.0 * sqrt(m))
jitter = np.random.uniform(-eps, eps, size=grid.shape)
grid_jitter = np.clip(grid + jitter, 0.0, 1.0)

x = grid_jitter[:, 0]
y = grid_jitter[:, 1]

plt.figure(figsize=(7, 5))
plt.scatter(x, y, color="steelblue", s=30, edgecolor="white", linewidth=0.5)
plt.title("MPG vs Weight")
plt.xlabel("Weight (1000 lb)")
plt.ylabel("Miles per Gallon (mpg)")
plt.tight_layout()
plt.show()

# %%
chat = chatlas.ChatAuto("openai/gpt-5")
chat.chat(
    "Interpret this plot.",
    chatlas.content_image_plot(),
)
